exporter/azuremonitorexporter/trace_to_envelope.go

```diff
@@ -0,0 +1,660 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azuremonitorexporter
+
+// Contains code common to both trace and metrics exporters
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Microsoft/ApplicationInsights-Go/appinsights/contracts"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+)
+
+const (
+	unknownSpanType   spanType = 0
+	httpSpanType      spanType = 1
+	rpcSpanType       spanType = 2
+	databaseSpanType  spanType = 3
+	messagingSpanType spanType = 4
+	faasSpanType      spanType = 5
+)
+
+var (
+	errUnspecifiedSpanKind          = errors.New("SpanKind is unspecified")
+	errUnexpectedAttributeValueType = errors.New("attribute value type is unexpected")
+)
+
+// Used to identify the type of a received Span
+type spanType int8
+
+// Transforms a tuple of pdata.Resource, pdata.InstrumentationLibrary, pdata.Span into an AppInsights contracts.Envelope
+// This is the only method that should be targeted in the unit tests
+func spanToEnvelope(
+	resource pdata.Resource,
+	instrumentationLibrary pdata.InstrumentationLibrary,
+	span pdata.Span,
+	logger *zap.Logger) (*contracts.Envelope, error) {
+
+	spanKind := span.Kind()
+
+	// unspecified, drop it
+	if spanKind == pdata.SpanKindUNSPECIFIED {
+		return nil, errUnspecifiedSpanKind
+	}
+
+	attributeMap := span.Attributes()
+	incomingSpanType := mapIncomingSpanToType(attributeMap)
+
+	envelope := contracts.NewEnvelope()
+	envelope.Tags = make(map[string]string)
+	envelope.Time = toTime(span.StartTime()).Format(time.RFC3339Nano)
+	traceIDHexString := idToHex(span.TraceID())
+	envelope.Tags[contracts.OperationId] = traceIDHexString
+	envelope.Tags[contracts.OperationParentId] = idToHex(span.ParentSpanID())
+
+	data := contracts.NewData()
+	var dataSanitizeFunc func() []string
+	var dataProperties map[string]string
+
+	if spanKind == pdata.SpanKindSERVER || spanKind == pdata.SpanKindCONSUMER {
+		requestData := spanToRequestData(span, incomingSpanType)
+		dataProperties = requestData.Properties
+		dataSanitizeFunc = requestData.Sanitize
+		envelope.Name = requestData.EnvelopeName("")
+		envelope.Tags[contracts.OperationName] = requestData.Name
+		data.BaseData = requestData
+		data.BaseType = requestData.BaseType()
+	} else if spanKind == pdata.SpanKindCLIENT || spanKind == pdata.SpanKindPRODUCER {
+		remoteDependencyData := spanToRemoteDependencyData(span, incomingSpanType)
+		dataProperties = remoteDependencyData.Properties
+		dataSanitizeFunc = remoteDependencyData.Sanitize
+		envelope.Name = remoteDependencyData.EnvelopeName("")
+		data.BaseData = remoteDependencyData
+		data.BaseType = remoteDependencyData.BaseType()
+	} else if spanKind == pdata.SpanKindINTERNAL {
+		messageData := spanToMessageData(span, incomingSpanType)
```

Comment:
I saw that both Azure Monitor exporter for Python (see [here](https://github.com/microsoft/opentelemetry-azure-monitor-python/blob/e9e669f51696d081f9816e4121f269a124fef6a2/azure_monitor/src/azure_monitor/export/trace/__init__.py#L143)) and JavaScript (see [here](https://github.com/microsoft/opentelemetry-azure-monitor-js/blob/295db7379cb504616be6aedd1983aee88e8bb7c6/packages/azure-opentelemetry-exporter/src/utils/spanUtils.ts#L193)) map internal spans to a remote dependency data with the type `InProc`. This made sense to me because message data doesn't have an ID and so it can't have children.

In my mind I've always mapped span events to message data instead.

However I am quite new to Azure Monitor, so I may be misunderstanding some things.

Reply:
An excellent point! @youngbupark suggested MessageData but I think I will change this back to InProc. I believe it was like that before.

Reply:
Good point. I also found https://docs.microsoft.com/en-us/azure/azure-monitor/app/opencensus-python-dependency#in-process-dependencies . The original code seems correct. apologies for wrong info.

Reply:
I totally agree with this and I'll make the change. I think it is InProc in the current version...